File: 290.word-pattern/solution.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        keys = {}
        values = {}
        words = s.split(" ")
        if (len(words) != len(pattern)):
            return False
        for i, pChar in enumerate(pattern):
            word = words[i]
            if word in values:
                if (values[word] != pChar):
                    return False
            elif pChar in keys:
                if (keys[pChar] != word):
                    return False
            else:
                keys[pChar] = word
                values[word] = pChar
        return True

File: 290.word-pattern/solution.kt
class Solution {
    fun wordPattern(pattern: String, s: String): Boolean {
        val words = s.split(" ")
        if (pattern.length != words.size) return false

        val patternToWord = HashMap<String, String>()
        val wordToPattern = HashMap<String, String>()

        for (i in 0 until words.size) {
            val word = words[i]
            val pChar = pattern[i].toString()
            if (wordToPattern.containsKey(word)) {
                if (pChar != wordToPattern.get(word)) return false

            } else if (patternToWord.containsKey(pChar)) {
                if (word != patternToWord.get(pChar)) return false

            } else {
                patternToWord.put(pChar, word)
                wordToPattern.put(word, pChar)
            }
        }
        return true
    }
}
